### PR TITLE
Remove Immaterial storage, ArrayView, and StreamView classes

### PR DESCRIPTION
**Pull request by tbennun**
- [x] Remove immaterial storage type
    - [x] Redo MPI tests without immaterial storage
- [x] Replace StreamArray with a pointer to a stream
- [x] Remove ArrayView
- [x] Revamp `defined_vars` to include actual variable types
- [x] Remove StreamView

### CODE REVIEW

File: dace/codegen/targets/fpga.py [hunk 1]
```diff
@@ -420,13 +423,14 @@ def allocate_array(self, sdfg, dfg, state_id, node, function_stream,
                                                               bank)
                         else:
                             self._bank_assignments[(dataname, sdfg)] = None
+                        # TODO: Remove auto keyword
```
Comment (definelicht) [inline dace/codegen/targets/fpga.py:426]: Are you sure? This returns an `hlslib::ocl::Buffer<T, AccessType>`, it's not like we are preventing the compiler from doing optimizations here.

File: dace/codegen/targets/intel_fpga.py [hunk 2]
```diff
@@ -176,6 +176,7 @@ def define_stream(self, dtype, buffer_size, var_name, array_size,
         kernel_stream.write("channel {} {}{}{};".format(vec_type, var_name,
                                                         size_str,
                                                         depth_attribute))
+        return 'channel {}'.format(vec_type)
```
Comment (definelicht) [inline dace/codegen/targets/intel_fpga.py:179]: Both a `kernel_stream.write` and a return in the same function?

File: dace/codegen/targets/xilinx.py [hunk 3]
```diff
@@ -186,20 +187,20 @@ def get_generated_codeobjects(self):
...
+                                                        var_name))
         else:
-            kernel_stream.write("dace::FIFO<{}, {}, {}> {}[{}];\n".format(
-                dtype.base_type.ctype, dtype.veclen, buffer_size, var_name,
-                cpp.sym2cpp(array_size)))
+            kernel_stream.write("{} {}[{}];\n".format(ctype, var_name,
+                                                      cpp.sym2cpp(array_size)))
             kernel_stream.write("dace::SetNames({}, \"{}\", {});".format(
                 var_name, var_name, cpp.sym2cpp(array_size)))
+        return ctype
```
Comment (definelicht) [inline dace/codegen/targets/xilinx.py:200]: Both write and return?

File: dace/runtime/include/dace/xilinx/stream.h [hunk 4]
```diff
@@ -14,12 +14,12 @@ template <typename T, unsigned vector_length, unsigned capacity>
 class FIFO {
  public:
   FIFO() : stream_(capacity) {
-    #pragma HLS INLINE
-    #pragma HLS STREAM variable=stream_ depth=capacity
+#pragma HLS INLINE
+#pragma HLS STREAM variable = stream_ depth = capacity
```
Comment (definelicht) [inline dace/runtime/include/dace/xilinx/stream.h:18]: You broke the pragma formatting :-) clang-format [cannot handle this](https://stackoverflow.com/questions/48639518/tell-clang-format-to-ignore-pragmas).

File: doc/source/dace.rst [hunk 5]
```diff
@@ -71,6 +71,14 @@ dace.serialize module
    :undoc-members:
    :show-inheritance:
 
+dace.sdfg module
```
Comment (definelicht) [inline doc/source/dace.rst:74]: What's happening here?

File: tests/codegen/mpi_axpy.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+import numpy as np
+import scipy as sp
+from mpi4py import MPI
+from dace.transformation.dataflow import MPITransformMap
+
+N = dace.symbol('N')
+
+
+@dace.program(dace.float64, dace.float64[N], dace.float64[N])
+def axpy(A, X, Y):
```
Comment (definelicht) [inline tests/codegen/mpi_axpy.py:14]: Is this related to this PR?

**Comment by definelicht** [comment]
I think you should add that immaterial is being removed to the PR title!

File: Jenkinsfile [hunk 7]
```diff
@@ -15,6 +15,7 @@ pipeline {
                     pip3 install --upgrade --user ".[testing]" .
                     pip3 install --user cmake
                     pip3 install --user coverage
+                    pip3 install --user mpi4py
```
[the remark this answers is not in the document]
Comment (tbennun) [reply Jenkinsfile:18]: Yes, see the bullet points above - I removed Immaterial (because they extensively use arrayviews) and thus had to rewrite the MPI tests.

File: dace/codegen/targets/cpu.py [hunk 8]
```diff
@@ -754,8 +804,34 @@ def process_out_memlets(self,
                                 dtype=node.out_connectors[uconn]) + ';', sdfg,
                             state_id, node)
                     else:
+                        defined_type, _ = self._dispatcher.defined_vars.get(
+                            memlet.data)
+                        if defined_type == DefinedType.Scalar:
+                            expr = memlet.data
+                        elif defined_type == DefinedType.ArrayInterface:
```
[the remark this answers is not in the document]
Comment (tbennun) [reply dace/codegen/targets/cpu.py:811]: Yes, this is planned to be completely revamped in the next two PRs (one function for all code generators that can be overridden)

File: tests/codegen/mpi_axpy.py (hunk 6 above)
Comment (tbennun) [reply tests/codegen/mpi_axpy.py:14]: Yes, see above comment about MPI tests being redone.

**Comment by tbennun** [comment]
@definelicht it's in the contents of the PR.

File: doc/source/dace.rst (hunk 5 above)
Comment (tbennun) [reply doc/source/dace.rst:74]: Documentation is missing for the SDFG module in the readthedocs website. It's not directly related to the PR.

File: dace/runtime/include/dace/xilinx/stream.h (hunk 4 above)
Comment (tbennun) [reply dace/runtime/include/dace/xilinx/stream.h:18]: oh no! are there any other files clang-format killed?

**Comment by definelicht** [comment]
> @definelicht it's in the contents of the PR.

I know - thanks for adding it to the title :-)

File: dace/runtime/include/dace/xilinx/stream.h (hunk 4 above)
Comment (definelicht) [reply dace/runtime/include/dace/xilinx/stream.h:18]: Only this one as far as I can see

Comment (tbennun) [reply dace/runtime/include/dace/xilinx/stream.h:18]: Fixed

File: dace/codegen/targets/fpga.py (hunk 1 above)
Comment (tbennun) [reply dace/codegen/targets/fpga.py:426]: Fixed

File: dace/codegen/targets/cpp.py [hunk 9]
```diff
@@ -661,104 +756,78 @@ def visit_Assign(self, node):
...
-        slice = self.visit(node.slice)
-        if not isinstance(slice, ast.Index):
-            raise NotImplementedError("Range subscripting not implemented")
+        subscript = self._subscript_expr(node.slice, target)
 
-        if isinstance(slice.value, ast.Tuple):
-            subscript = unparse(slice)[1:-1]
-        else:
-            subscript = unparse(slice)
+        # Creating as a malformed ast.Name object so it does not visit again
```
[the remark this answers is not in the document]
Comment (tbennun) [reply dace/codegen/targets/cpp.py:828]: Done.

File: dace/codegen/targets/intel_fpga.py (hunk 2 above)
Comment (tbennun) [reply dace/codegen/targets/intel_fpga.py:179]: The return value is for `defined_vars` rather than to generate code. Otherwise we have to copy defined_vars everywhere.

File: dace/codegen/targets/xilinx.py (hunk 3 above)
Comment (tbennun) [reply dace/codegen/targets/xilinx.py:202]: See above. Now fixed.

File: dace/runtime/include/dace/xilinx/stream.h (hunk 4 above)
Comment (tbennun) [reply dace/runtime/include/dace/xilinx/stream.h]: if you refresh and look at the diff, it should be gone (I double checked)
